pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -669,6 +670,106 @@ type TestResource struct {
 	topology         bool
 }
 
+func TestFilterByAffinity(t *testing.T) {
+	as := require.New(t)
+	allDevices := ResourceDeviceInstances{
+		"res1": map[string]pluginapi.Device{
+			"dev1": {
+				ID: "dev1",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 1,
+						},
+					},
+				},
+			},
+			"dev2": {
+				ID: "dev2",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 1,
+						},
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+			"dev3": {
+				ID: "dev3",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+			"dev4": {
+				ID: "dev4",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testManager := ManagerImpl{
+		topologyAffinityStore: newFakeTopologyManager(),
+		allDevices:            allDevices,
+	}
+
+	testCases := []struct {
+		available               sets.String
+		fromAffinityExpected    sets.String
+		notFromAffinityExpected sets.String
+		withoutTopologyExpected sets.String
+	}{
+		{
+			available:               sets.NewString("dev1", "dev2"),
+			fromAffinityExpected:    sets.NewString("dev2"),
+			notFromAffinityExpected: sets.NewString("dev1"),
+			withoutTopologyExpected: sets.NewString(),
+		},
+		{
+			available:               sets.NewString("dev1", "dev2", "dev3", "dev4"),
+			fromAffinityExpected:    sets.NewString("dev2", "dev3", "dev4"),
+			notFromAffinityExpected: sets.NewString("dev1"),
+			withoutTopologyExpected: sets.NewString(),
+		},
+	}
+
+	for _, testCase := range testCases {
+		fromAffinity, notFromAffinity, withoutTopology := testManager.filterByAffinity("", "", "res1", testCase.available)
```

Comment:
Not a fan of this: in general we should always test public interfaces (and fix the interface accordingly if it is too hard), but in this case it makes sense.

Reply:
aha, I think we can do this by extending code in TestPodContainerDeviceAllocation

Reply:
If you have the chance please do. Reusing/extending existing testing infra (which should also something @klueska suggested IIUC) and testing against the public interface is something that almost always is a good direction.

Reply:
@fromanirh  I tried but found it too hard to reproduce the problem steadily by testing the public `Allocate()`interface  like in `TestPodContainerDeviceAllocation()`, the problem is that we are using sets.String{} to store the filter results, and the sequence produced by a loop of sets.String{} is random:

https://github.com/kubernetes/kubernetes/blob/7ccd90e7d7c6d727a7497c901c45776592ec5e45/pkg/kubelet/cm/devicemanager/manager.go#L675-L685

also`devicesToAllocate()` will collect result from all the filter result: aligned, unaligned, noAffinity: 

https://github.com/kubernetes/kubernetes/blob/7ccd90e7d7c6d727a7497c901c45776592ec5e45/pkg/kubelet/cm/devicemanager/manager.go#L706-L750

so the result is nondeterministic, I got 74 failed results when I run the single test for 1024 times. In another way, making `filterByAffinity()` a public interface does not seems like a good idea, any better suggestion?

Reply:
in this case testing the private method, lacking better approaches, is OK with me.
In general I'd like to review later on if we need to use `sets.String` or if a simple ordered `[]string` could suffice and enable better testing. Not blocking for this PR anyway.